File: phpmailer-go/smtp.go
package phpmailer

import (
	"bufio"
	"fmt"
	"net"
	"strings"
	"time"
)

type SMTPSecurity string

const (
	SMTPSecurityNone     SMTPSecurity = "none"
	SMTPSecurityStartTLS SMTPSecurity = "starttls"
	SMTPSecuritySSL      SMTPSecurity = "ssl"
)

type SMTPConnection struct {
	Host           string
	Port           int
	TimeoutSeconds int
	Security       SMTPSecurity
	Conn           net.Conn
	ServerHello    string
}

func NewSMTPConnection(host string, port int) *SMTPConnection {
	return &SMTPConnection{
		Host:           host,
		Port:           port,
		TimeoutSeconds: 300,
		Security:       SMTPSecurityNone,
	}
}

func (s *SMTPConnection) Connect() error {
	address := fmt.Sprintf("%s:%d", s.Host, s.Port)
	conn, err := net.DialTimeout("tcp", address, time.Duration(s.TimeoutSeconds)*time.Second)
	if err != nil {
		return wrapMailError(ErrorTransport, "failed to connect to smtp", err)
	}

	s.Conn = conn
	s.ServerHello, err = s.ReadResponse()
	if err != nil {
		return err
	}

	if !strings.HasPrefix(s.ServerHello, "220") {
		return newMailError(ErrorProtocol, "unexpected smtp greeting: "+s.ServerHello)
	}

	return nil
}

func (s *SMTPConnection) SendCommand(command string) (string, error) {
	if s.Conn == nil {
		return "", newMailError(ErrorTransport, "smtp connection is nil")
	}

	_, err := s.Conn.Write([]byte(command + "\r\n"))
	if err != nil {
		return "", wrapMailError(ErrorTransport, "smtp command write failed", err)
	}

	response, err := s.ReadResponse()
	if err != nil {
		return "", err
	}

	return response, nil
}

func (s *SMTPConnection) ReadResponse() (string, error) {
	if s.Conn == nil {
		return "", newMailError(ErrorTransport, "smtp connection is nil")
	}

	deadline := time.Now().Add(time.Duration(s.TimeoutSeconds) * time.Second)
	_ = s.Conn.SetReadDeadline(deadline)

	reader := bufio.NewReader(s.Conn)
	lines := make([]string, 0)

	for {
		line, err := reader.ReadString('\n')
		if err != nil {
			return "", wrapMailError(ErrorTransport, "smtp read failed", err)
		}

		cleanLine := strings.TrimRight(line, "\r\n")
		lines = append(lines, cleanLine)

		if len(cleanLine) >= 4 && cleanLine[3] == ' ' {
			break
		}
	}

	return strings.Join(lines, "\n"), nil
}

func (s *SMTPConnection) Quit() error {
	if s.Conn == nil {
		return nil
	}

	_, _ = s.SendCommand("QUIT")
	err := s.Conn.Close()
	s.Conn = nil
	if err != nil {
		return wrapMailError(ErrorTransport, "smtp close failed", err)
	}

	return nil
}

File: phpmailer-rs/src/smtp.rs
use std::io::{BufRead, BufReader, Write};
use std::net::TcpStream;
use std::time::Duration;

use crate::phpmailer::MailError;

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum SmtpSecurity {
    None,
    StartTls,
    Ssl,
}

#[derive(Debug)]
pub struct SmtpConnection {
    pub host: String,
    pub port: u16,
    pub timeout_seconds: u64,
    pub security: SmtpSecurity,
    pub stream: Option<TcpStream>,
    pub server_hello: String,
}

impl Clone for SmtpConnection {
    fn clone(&self) -> Self {
        Self {
            host: self.host.clone(),
            port: self.port,
            timeout_seconds: self.timeout_seconds,
            security: self.security.clone(),
            stream: None,
            server_hello: self.server_hello.clone(),
        }
    }
}

impl SmtpConnection {
    pub fn new(host: String, port: u16) -> Self {
        Self {
            host: host.clone(),
            port,
            timeout_seconds: 300,
            security: SmtpSecurity::None,
            stream: None,
            server_hello: String::new(),
        }
    }

    pub fn connect(&mut self) -> Result<(), MailError> {
        let address = format!("{}:{}", self.host.clone(), self.port);
        let stream = TcpStream::connect(address.clone())
            .map_err(|err| MailError::transport(format!("Failed to connect to {address}: {err}")))?;

        stream
            .set_read_timeout(Some(Duration::from_secs(self.timeout_seconds)))
            .map_err(|err| MailError::transport(format!("Failed to set read timeout: {err}")))?;

        stream
            .set_write_timeout(Some(Duration::from_secs(self.timeout_seconds)))
            .map_err(|err| MailError::transport(format!("Failed to set write timeout: {err}")))?;

        self.stream = Some(stream);
        self.server_hello = self.read_response()?;

        if !self.server_hello.starts_with("220") {
            return Err(MailError::protocol(format!(
                "Unexpected SMTP greeting: {}",
                self.server_hello.clone()
            )));
        }

        Ok(())
    }

    pub fn send_command(&mut self, command: String) -> Result<String, MailError> {
        let stream = self
            .stream
            .as_mut()
            .ok_or_else(|| MailError::transport("SMTP stream is not connected".to_string()))?;

        let payload = format!("{}\r\n", command.clone());
        stream
            .write_all(payload.as_bytes())
            .map_err(|err| MailError::transport(format!("Write command failed ({command}): {err}")))?;

        stream
            .flush()
            .map_err(|err| MailError::transport(format!("Flush command failed ({command}): {err}")))?;

        self.read_response()
    }

    pub fn read_response(&mut self) -> Result<String, MailError> {
        let stream = self
            .stream
            .as_mut()
            .ok_or_else(|| MailError::transport("SMTP stream is not connected".to_string()))?;

        let mut reader = BufReader::new(stream);
        let mut response = String::new();
        let mut line = String::new();

        loop {
            line.clear();
            let bytes = reader
                .read_line(&mut line)
                .map_err(|err| MailError::transport(format!("Read response failed: {err}")))?;

            if bytes == 0 {
                break;
            }

            response.push_str(line.clone().as_str());

            let is_last_line = line.as_bytes().get(3).copied() == Some(b' ');
            if line.len() >= 4 && is_last_line {
                break;
            }
        }

        Ok(response.trim_end().to_string())
    }

    pub fn quit(&mut self) -> Result<(), MailError> {
        if self.stream.is_none() {
            return Ok(());
        }

        let _ = self.send_command("QUIT".to_string());
        self.stream = None;

        Ok(())
    }
}
